lib/common/src/Amp.php

```diff
@@ -0,0 +1,261 @@
+     *
+     * The Amp viewer script node is of the form '<script async
+     * src="https://cdn.ampproject.org/v0/amp-viewer-integration-...js>"</script>'.
+     *
+     * @param DOMNode $node Node to check.
+     * @return bool Whether the given node is the Amp runtime script.
+     */
+    public static function isViewerScript(DOMNode $node)
+    {
+        if (
+            ! self::isAsyncScript($node)
+            || self::isExtension($node)
+        ) {
+            return false;
+        }
+
+        $src = $node->getAttribute(Attribute::SRC);
+
+        if (strpos($src, self::CACHE_HOST . '/v0/amp-viewer-integration-') !== 0) {
+            return false;
+        }
+
+        if (substr($src, -3) !== '.js') {
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Check if a given node is an Amp extension.
+     *
+     * @param DOMNode $node Node to check.
+     * @return bool Whether the given node is the Amp runtime script.
+     */
+    public static function isExtension(DOMNode $node)
+    {
+        return ! empty(self::getExtensionName($node));
+    }
+
+    /**
+     * Get the name of the extension.
+     *
+     * Returns an empty string if the name of the extension could not be retrieved.
+     *
+     * @param DOMNode $node Node to get the name of.
+     * @return string Name of the custom node or template. Empty string if none found.
+     */
+    public static function getExtensionName(DOMNode $node)
+    {
+        if (! $node instanceof DOMElement || $node->tagName !== Tag::SCRIPT) {
+            return '';
+        }
+
+        if ($node->hasAttribute(Attribute::CUSTOM_ELEMENT)) {
+            return $node->getAttribute(Attribute::CUSTOM_ELEMENT);
+        }
+
+        if ($node->hasAttribute(Attribute::CUSTOM_TEMPLATE)) {
+            return $node->getAttribute(Attribute::CUSTOM_TEMPLATE);
+        }
+
+        if ($node->hasAttribute(Attribute::HOST_SERVICE)) {
+            return $node->getAttribute(Attribute::HOST_SERVICE);
+        }
+
+        return '';
+    }
+
+    /**
+     * Check whether a given node is a script for a render-delaying extension.
+     *
+     * @param DOMNode $node Node to check.
+     * @return bool Whether the node is a script for a render-delaying extension.
+     */
+    public static function isRenderDelayingExtension(DOMNode $node)
+    {
+        $extensionName = self::getExtensionName($node);
+
+        if (empty($extensionName)) {
+            return false;
+        }
+
+        return in_array($extensionName, self::RENDER_DELAYING_EXTENSIONS, true);
+    }
+
+    /**
+     * Check whether a given DOM node is an Amp custom element.
+     *
+     * @param DOMNode $node DOM node to check.
+     * @return bool Whether the checked DOM node is an Amp custom element.
+     */
+    public static function isCustomElement(DOMNode $node)
+    {
+        return $node instanceof DOMElement && strpos($node->tagName, Extension::PREFIX) === 0;
```

Comment:
```suggestion
        return $node instanceof DOMElement && Extension::PREFIX === substr($node->tagName, 0, strlen(Extension::PREFIX));
```

Reply:
Again, I think this makes it less maintainable and less performant. You one operation into two operations, and duplicate the places you need to adapt in case of a change (constant referenced twice).